lib/role_picker.rb

```diff
@@ -1,35 +1,32 @@
 class RolePicker
-  attr_accessor :user
-  GUARDS = {
-    requester: ->(u, p) { u == p.requester },
-    active_approver: ->(u, p) { p.is_active_approver? u },
-    observer: ->(u, p) { p.observers.include? u },
-    approver: ->(u, p) { p.approvers.include? u },
-    client_admin: ->(u, p) { u.client_admin? && p.client_slug == u.client_slug }
-  }
+  attr_reader :user
 
   def initialize(user, proposal)
     @user = user
-    @role_types = []
-    GUARDS.each do |role, guard|
-      if guard[user, proposal]
-        @role_types << role
-      end
-    end
+    @proposal = proposal
   end
 
-  # Active Observer means, their main role is an observer
   def active_observer?
     observer? && !active_approver? && !requester?
   end
 
-  # observer? approver? etc. get converted into scanning role_types
-  def method_missing(method_sym, *arguments, &block)
-    method_str = method_sym.to_s
-    if method_str.end_with?("?")
-      @role_types.include?(method_str[0..-2].to_sym)
-    else
-      super
-    end
+  def active_approver?
+    proposal.is_active_approver?(user)
   end
+
+  def requester?
+    user == proposal.requester
+  end
+
+  def observer?
+    proposal.observers.include?(user)
+  end
+
+  def approver?
+    proposal.approvers.include?(user)
+  end
+
+  private
+
+  attr_reader :proposal
 end
```

Comment:
I like that you've split that auto-generating code up into explicit method definitions. Much easier to grok for first-timer.